Hercules crashes with language filter

Hi!

I tried hercules on a quite big private repository of our company. It crashes when I use the language filter. If we look at the CLI output below we see that the first variant without the `--languages csharp` runs and generate valid (and interesting, thanks!) data. When I add the language filter it crashes after some time.

```
$ ./hercules.linux_amd64 --burndown --granularity 2 --sampling 2 --pb --commits develop_stretching_hashes repo-cache > baukasten_burndown_stretching.pb
./hercules.linux_amd64 --burndown --granularity 2 --sampling 2 --pb --commits  8785,81s user 60,45s system 103% cpu 2:22:28,83 total

$ ./hercules.linux_amd64 --burndown --granularity 2 --sampling 2 --pb --commits develop_stretching_hashes repo-cache --languages csharp > baukasten_burndown_stretching_csharp_only.pb
finalizing...2019/04/07 20:23:49 Failed to run the pipeline on [git@github.com:ORG/REPO.git]
panic: empty history

goroutine 1 [running]:
gopkg.in/src-d/hercules.v10/leaves.(*BurndownAnalysis).groupSparseHistory(0xc002248400, 0xc002eddf50, 0xffffffffffffffff, 0xc02947b458, 0x88043c, 0xc000000180, 0x300000002)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/leaves/burndown.go:1472 +0x703
gopkg.in/src-d/hercules.v10/leaves.(*BurndownAnalysis).Finalize(0xc002248400, 0x15d2540, 0xc0000dc100)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/leaves/burndown.go:504 +0x55
gopkg.in/src-d/hercules.v10/internal/core.(*Pipeline).Run(0xc02947bcc8, 0xc003736000, 0x12b8, 0x1400, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/internal/core/pipeline.go:866 +0x714
main.glob..func3(0x21d3ec0, 0xc000a6a0b0, 0x1, 0xb)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/cmd/hercules/root.go:270 +0x85b
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).execute(0x21d3ec0, 0xc0000cc010, 0xb, 0xb, 0x21d3ec0, 0xc0000cc010)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:766 +0x2ae
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x21d3ec0, 0xc000a225e0, 0xc00016ff88, 0x84f65f)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:852 +0x2ec
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:800
main.main()
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/cmd/hercules/root.go:510 +0x32
./hercules.linux_amd64 --burndown --granularity 2 --sampling 2 --pb --commits  854,87s user 46,41s system 108% cpu 13:51,67 total
```
As I can not give anyone access to the repository to reproduce this error feel free to close this issue if you want. I will try in the meantime to reduce the commandline flags used to produce the crash. This can take some time because the crash happend after 1/4 hour.
And yes, there are C# files :D.

Edit: More crashes
```
./hercules.linux_amd64 --burndown --granularity 2 --sampling 2 --pb repo-cache --languages csharp > baukasten_burndown_csharp_only.pb                                               
finalizing...2019/04/07 20:38:33 Failed to run the pipeline on [git@github.com:ORG/REPO.git]
panic: empty history

goroutine 1 [running]:
gopkg.in/src-d/hercules.v10/leaves.(*BurndownAnalysis).groupSparseHistory(0xc0087c0f00, 0xc002542cf0, 0xffffffffffffffff, 0xc01b679458, 0x88043c, 0xc000000180, 0x300000002)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/leaves/burndown.go:1472 +0x703
gopkg.in/src-d/hercules.v10/leaves.(*BurndownAnalysis).Finalize(0xc0087c0f00, 0x15d2540, 0xc0000da300)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/leaves/burndown.go:504 +0x55
gopkg.in/src-d/hercules.v10/internal/core.(*Pipeline).Run(0xc01b679cc8, 0xc002bb4000, 0x8b9, 0x900, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/internal/core/pipeline.go:866 +0x714
main.glob..func3(0x21d3ec0, 0xc000a623f0, 0x1, 0x9)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/cmd/hercules/root.go:270 +0x85b
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).execute(0x21d3ec0, 0xc0000cc010, 0x9, 0x9, 0x21d3ec0, 0xc0000cc010)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:766 +0x2ae
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x21d3ec0, 0xc000a26620, 0xc00016ff88, 0x84f65f)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:852 +0x2ec
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:800
main.main()
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/cmd/hercules/root.go:510 +0x32
./hercules.linux_amd64 --burndown --granularity 2 --sampling 2 --pb repo-cach  475,91s user 19,12s system 105% cpu 7:49,19 total

$ ./hercules.linux_amd64 --burndown --pb repo-cache --languages csharp > baukasten_burndown_csharp_only.pb                             
finalizing...2019/04/07 20:46:45 Failed to run the pipeline on [git@github.com:ORG/REPO.git]
panic: empty history

goroutine 1 [running]:
gopkg.in/src-d/hercules.v10/leaves.(*BurndownAnalysis).groupSparseHistory(0xc00de68e00, 0xc00329cba0, 0xffffffffffffffff, 0xc007ee7458, 0x88043c, 0xc000000180, 0x300000002)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/leaves/burndown.go:1472 +0x703
gopkg.in/src-d/hercules.v10/leaves.(*BurndownAnalysis).Finalize(0xc00de68e00, 0x15d2540, 0xc0000b2300)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/leaves/burndown.go:504 +0x55
gopkg.in/src-d/hercules.v10/internal/core.(*Pipeline).Run(0xc007ee7cc8, 0xc002144000, 0x8b9, 0x900, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/internal/core/pipeline.go:866 +0x714
main.glob..func3(0x21d3ec0, 0xc000a3a550, 0x1, 0x5)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/cmd/hercules/root.go:270 +0x85b
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).execute(0x21d3ec0, 0xc0000321f0, 0x5, 0x5, 0x21d3ec0, 0xc0000321f0)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:766 +0x2ae
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x21d3ec0, 0xc000a1e600, 0xc00013ff88, 0x84f65f)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:852 +0x2ec
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:800
main.main()
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/cmd/hercules/root.go:510 +0x32

$ ./hercules.linux_amd64 --burndown --languages csharp repo-cache
finalizing...2019/04/07 20:50:36 Failed to run the pipeline on [git@github.com:ORG/REPO.git]
panic: empty history

goroutine 1 [running]:
gopkg.in/src-d/hercules.v10/leaves.(*BurndownAnalysis).groupSparseHistory(0xc0000c1a00, 0xc00245ba40, 0xffffffffffffffff, 0xc0131e7458, 0x88043c, 0xc000000180, 0x300000002)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/leaves/burndown.go:1472 +0x703
gopkg.in/src-d/hercules.v10/leaves.(*BurndownAnalysis).Finalize(0xc0000c1a00, 0x15d2540, 0xc0000c0100)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/leaves/burndown.go:504 +0x55
gopkg.in/src-d/hercules.v10/internal/core.(*Pipeline).Run(0xc0131e7cc8, 0xc0027f6000, 0x8b9, 0x900, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/internal/core/pipeline.go:866 +0x714
main.glob..func3(0x21d3ec0, 0xc000a3e200, 0x1, 0x4)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/cmd/hercules/root.go:270 +0x85b
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).execute(0x21d3ec0, 0xc000032060, 0x4, 0x4, 0x21d3ec0, 0xc000032060)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:766 +0x2ae
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x21d3ec0, 0xc000233690, 0xc00014df88, 0x84f65f)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:852 +0x2ec
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:800
main.main()
	/home/travis/gopath/src/gopkg.in/src-d/hercules.v10/cmd/hercules/root.go:510 +0x32
```

# Versions
```
 $ pip freeze | rg labours
labours==10.0.1

 $ ./hercules.linux_amd64 version
Version: 10
Git:     b856b666909194669e93d2c8dd5f86a96d9f60dc
 ```

Just to confirm, same here with `--languages python`:

```sh
$ ../hercules.linux_amd64 --burndown --burndown-people --languages python . > ../out.yaml
```

It works till the very end and then crashes with 'empty history' (it's definitely not empty...):

```
finalizing...2019/04/08 12:06:22 Failed to run the pipeline on [...]
panic: empty history

goroutine 1 [running]:
gopkg.in/src-d/hercules.v10/leaves.(*BurndownAnalysis).groupSparseHistory(0xc000b0ee00, 0xc00247eb40, 0xffffffffffffffff, 0xc000048070, 0xc000048000, 0x0, 0x91d511)
        /home/travis/gopath/src/gopkg.in/src-d/hercules.v10/leaves/burndown.go:1472 +0x715
gopkg.in/src-d/hercules.v10/leaves.(*BurndownAnalysis).Finalize(0xc000b0ee00, 0x1535660, 0xc000b0e000)
        /home/travis/gopath/src/gopkg.in/src-d/hercules.v10/leaves/burndown.go:504 +0x55
gopkg.in/src-d/hercules.v10/internal/core.(*Pipeline).Run(0xc001a11c20, 0xc0024da000, 0x1bb5, 0x1c00, 0x0, 0x0, 0x0)
        /home/travis/gopath/src/gopkg.in/src-d/hercules.v10/internal/core/pipeline.go:852 +0x6cf
main.glob..func3(0x20e7d80, 0xc000b10550, 0x1, 0x5)
        /home/travis/gopath/src/gopkg.in/src-d/hercules.v10/cmd/hercules/root.go:270 +0x843
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).execute(0x20e7d80, 0xc0000301f0, 0x5, 0x5, 0x20e7d80, 0xc0000301f0)
        /home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:766 +0x2cc
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x20e7d80, 0xc000ae65c0, 0xc000ae6590, 0xc000b0e600)
        /home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:852 +0x2fd
gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra.(*Command).Execute(0x20e7d80, 0x84dd40, 0xc0000440b8)
        /home/travis/gopath/src/gopkg.in/src-d/hercules.v10/vendor/github.com/spf13/cobra/command.go:800 +0x2b
main.main()
        /home/travis/gopath/src/gopkg.in/src-d/hercules.v10/cmd/hercules/root.go:510 +0x31
```

--- a/internal/plumbing/tree_diff.go
+++ b/internal/plumbing/tree_diff.go
@@ -23,7 +23,9 @@ type TreeDiff struct {
 	core.NoopMerger
 	SkipFiles  []string
 	NameFilter *regexp.Regexp
-	Languages  map[string]bool
+	// Languages is the set of allowed languages. The values must be lower case. The default
+	// (empty) set disables the language filter.
+	Languages map[string]bool
 
 	previousTree   *object.Tree
 	previousCommit plumbing.Hash
@@ -130,7 +132,7 @@ func (treediff *TreeDiff) Configure(facts map[string]interface{}) error {
 	if val, exists := facts[ConfigTreeDiffLanguages].([]string); exists {
 		treediff.Languages = map[string]bool{}
 		for _, lang := range val {
-			treediff.Languages[strings.TrimSpace(lang)] = true
+			treediff.Languages[strings.ToLower(strings.TrimSpace(lang))] = true
 		}
 	} else if treediff.Languages == nil {
 		treediff.Languages = map[string]bool{}
@@ -281,7 +283,7 @@ func (treediff *TreeDiff) checkLanguage(name string, blobHash plumbing.Hash) (bo
 	if n < len(buffer) {
 		buffer = buffer[:n]
 	}
-	lang := enry.GetLanguage(path.Base(name), buffer)
+	lang := strings.ToLower(enry.GetLanguage(path.Base(name), buffer))
 	return treediff.Languages[lang], nil
 }
 

--- a/internal/plumbing/tree_diff_test.go
+++ b/internal/plumbing/tree_diff_test.go
@@ -78,7 +78,7 @@ func TestTreeDiffConsume(t *testing.T) {
 		"fbe766ffdc3f87f6affddc051c6f8b419beea6a2"))
 	td.previousTree, _ = prevCommit.Tree()
 	res, err := td.Consume(deps)
-	assert.Nil(t, err)
+	assert.NoError(t, err)
 	assert.Equal(t, len(res), 1)
 	changes := res[DependencyTreeChanges].(object.Changes)
 	assert.Equal(t, len(changes), 12)
@@ -98,7 +98,7 @@ func TestTreeDiffConsume(t *testing.T) {
 	}
 	for _, change := range changes {
 		action, err := change.Action()
-		assert.Nil(t, err)
+		assert.NoError(t, err)
 		if change.From.Name != "" {
 			assert.Contains(t, baseline, change.From.Name)
 			assert.Equal(t, baseline[change.From.Name], action)
@@ -116,13 +116,13 @@ func TestTreeDiffConsumeFirst(t *testing.T) {
 	deps := map[string]interface{}{}
 	deps[core.DependencyCommit] = commit
 	res, err := td.Consume(deps)
-	assert.Nil(t, err)
+	assert.NoError(t, err)
 	assert.Equal(t, len(res), 1)
 	changes := res[DependencyTreeChanges].(object.Changes)
 	assert.Equal(t, len(changes), 21)
 	for _, change := range changes {
 		action, err := change.Action()
-		assert.Nil(t, err)
+		assert.NoError(t, err)
 		assert.Equal(t, action, merkletrie.Insert)
 	}
 }
@@ -151,7 +151,7 @@ func TestTreeDiffConsumeSkip(t *testing.T) {
 		"1e076dc56989bc6aa1ef5f55901696e9e01423d4"))
 	td.previousTree, _ = prevCommit.Tree()
 	res, err := td.Consume(deps)
-	assert.Nil(t, err)
+	assert.NoError(t, err)
 	assert.Equal(t, len(res), 1)
 	changes := res[DependencyTreeChanges].(object.Changes)
 	assert.Equal(t, 37, len(changes))
@@ -164,7 +164,7 @@ func TestTreeDiffConsumeSkip(t *testing.T) {
 		ConfigTreeDiffBlacklistedPrefixes: []string{"vendor/"},
 	})
 	res, err = td.Consume(deps)
-	assert.Nil(t, err)
+	assert.NoError(t, err)
 	assert.Equal(t, len(res), 1)
 	changes = res[DependencyTreeChanges].(object.Changes)
 	assert.Equal(t, 31, len(changes))
@@ -182,7 +182,7 @@ func TestTreeDiffConsumeOnlyFilesThatMatchFilter(t *testing.T) {
 		"1e076dc56989bc6aa1ef5f55901696e9e01423d4"))
 	td.previousTree, _ = prevCommit.Tree()
 	res, err := td.Consume(deps)
-	assert.Nil(t, err)
+	assert.NoError(t, err)
 	assert.Equal(t, len(res), 1)
 	changes := res[DependencyTreeChanges].(object.Changes)
 	assert.Equal(t, 37, len(changes))
@@ -194,7 +194,7 @@ func TestTreeDiffConsumeOnlyFilesThatMatchFilter(t *testing.T) {
 		ConfigTreeDiffFilterRegexp: ".*go",
 	})
 	res, err = td.Consume(deps)
-	assert.Nil(t, err)
+	assert.NoError(t, err)
 	assert.Equal(t, len(res), 1)
 	changes = res[DependencyTreeChanges].(object.Changes)
 	assert.Equal(t, 27, len(changes))
@@ -208,7 +208,7 @@ func TestTreeDiffConsumeLanguageFilterFirst(t *testing.T) {
 	deps := map[string]interface{}{}
 	deps[core.DependencyCommit] = commit
 	res, err := td.Consume(deps)
-	assert.Nil(t, err)
+	assert.NoError(t, err)
 	assert.Equal(t, len(res), 1)
 	changes := res[DependencyTreeChanges].(object.Changes)
 	assert.Equal(t, len(changes), 6)
@@ -222,21 +222,29 @@ func TestTreeDiffConsumeLanguageFilterFirst(t *testing.T) {
 
 func TestTreeDiffConsumeLanguageFilter(t *testing.T) {
 	td := fixtureTreeDiff()
-	td.Configure(map[string]interface{}{ConfigTreeDiffLanguages: []string{"Python"}})
+	assert.NoError(t, td.Configure(map[string]interface{}{ConfigTreeDiffLanguages: []string{"Python"}}))
 	commit, _ := test.Repository.CommitObject(plumbing.NewHash(
 		"e89c1d10fb31e32668ad905eb59dc44d7a4a021e"))
 	deps := map[string]interface{}{}
 	deps[core.DependencyCommit] = commit
 	res, err := td.Consume(deps)
-	assert.Nil(t, err)
+	assert.NoError(t, err)
 	assert.Equal(t, len(res), 1)
 	commit, _ = test.Repository.CommitObject(plumbing.NewHash(
 		"fbe766ffdc3f87f6affddc051c6f8b419beea6a2"))
 	deps[core.DependencyCommit] = commit
 	res, err = td.Consume(deps)
-	assert.Nil(t, err)
+	assert.NoError(t, err)
 	assert.Equal(t, len(res), 1)
 	changes := res[DependencyTreeChanges].(object.Changes)
+	assert.Equal(t, len(changes), 1)
+	assert.Equal(t, changes[0].To.Name, "labours.py")
+
+	// lowercase
+	assert.NoError(t, td.Configure(map[string]interface{}{ConfigTreeDiffLanguages: []string{"python"}}))
+	assert.NoError(t, err)
+	assert.Equal(t, len(res), 1)
+	changes = res[DependencyTreeChanges].(object.Changes)
 	assert.Equal(t, len(changes), 1)
 	assert.Equal(t, changes[0].To.Name, "labours.py")
 }
@@ -258,7 +266,7 @@ func TestTreeDiffCheckLanguage(t *testing.T) {
 	td.Languages["Go"] = true
 	lang, err := td.checkLanguage(
 		"version.go", plumbing.NewHash("975f35a1412b8ae79b5ba2558f71f41e707fd5a9"))
-	assert.Nil(t, err)
+	assert.NoError(t, err)
 	assert.True(t, lang)
 }
 

